ui/src/views/Releases/ListReleaseRevisionsV2/index.jsx

```diff
@@ -0,0 +1,258 @@
+  // Balrog URL that will retrieve the current version of the
+  // entire Release rather than a GCS link that only contains
+  // the most recently changed part.
+  const currentRevision =
+    latestRevisions.length > 0
+      ? {
+          path: null,
+          timestamp: latestRevisions[0].timestamp,
+          changed_by: latestRevisions[0].changed_by,
+          data_url: `/v2/releases/${releaseName}`,
+        }
+      : {};
+  const prettyPath =
+    // eslint-disable-next-line no-nested-ternary
+    olderRevisions.length > 0
+      ? olderRevisions[leftRadioCheckedIndex - 1].path
+        ? olderRevisions[leftRadioCheckedIndex - 1].path
+        : 'Base'
+      : '';
+
+  useEffect(() => {
+    fetchRevisions(releaseName, 2);
+  }, [releaseName]);
+
+  const handleLeftRadioChange = ({ target: { value } }) => {
+    setLeftRadioCheckedIndex(Number(value));
+  };
+
+  const handleRightRadioChange = ({ target: { value } }) => {
+    setRightRadioCheckedIndex(Number(value));
+  };
+
+  const handleDrawerClose = () => {
+    setDrawerState({
+      ...drawerState,
+      open: false,
+    });
+  };
+
+  const handleViewClick = item => async () => {
+    const section =
+      // eslint-disable-next-line no-nested-ternary
+      item.path === null
+        ? 'entire Release'
+        : item.path === ''
+        ? 'base'
+        : item.path;
+    const when = new Date(item.timestamp);
+    const result = await fetchRevisionData(item.data_url);
+
+    setDrawerState({
+      ...drawerState,
+      // Result could come from Balrog, where what we want to display is in
+      // `blob` or from GCS, where its the entire `data`.
+      item: result.data.data.blob ? result.data.data.blob : result.data.data,
+      title: `Contents of ${section} at ${when}`,
+      open: true,
+    });
+  };
+
+  useEffect(() => {
+    const olderRevision = olderRevisions[leftRadioCheckedIndex - 1];
+
+    if (olderRevision) {
+      // The left revision is always a part of the Release
+      // that we can retrieve from GCS.
+      axios.get(olderRevision.data_url).then(result => {
+        setLeftRevisionData(result.data || {});
+      });
+
+      // In order to display a useful diff, we need the right revision
+      // to be the most recent version of the same part of the Release,
+      // which we can find by looking for the older revisions path in
+      // `latestRevisions`.
+      const latestRevision = latestRevisions.find(
+        r => r.path === olderRevision.path
+      );
+
+      if (latestRevision) {
+        axios.get(latestRevision.data_url).then(result => {
+          // FIXME: sometimes the page doesn't rerender the diff after this
+          // for some reason...
+          // I think there's some sort of race condition, or issue with the
+          // depedencies on this useEffect.
+          setRightRevisionData(result.data || {});
+        });
+      }
+    }
+  }, [leftRadioCheckedIndex, fetchedRevisions.data]);
```

Comment:
`olderRevisions`, `latestRevisions` are variables defined outside `useEffect` but are not listed in the list of dependencies in `useEffect`. Maybe that would fix the issue you were seeing?

Reply:
I'll give that a shot!

Reply:
That was definitely part of the problem. After some further digging, I think this is actually caused by `DiffRelease`. Specifically, anytime any of the dependencies in https://github.com/mozilla-releng/balrog/blob/master/ui/src/components/DiffRelease/index.jsx#L87 change, it fires a new worker. When I change the left radio button, I see 3 workers run (one for `firstFilename`, one for `firstRelease`, and one for `secondRelease`) - and the final one ends up winning.

I'm guessing that we're getting a rerender of that component each time one of these changes, although I'm not sure how to prevent that.

Reply:
What do you think about the following. I think it might prevent the rerendering.

```diff
diff --git a/ui/src/components/DiffRelease/diff.worker.js b/ui/src/components/DiffRelease/diff.worker.js
index 4192ce8c..59e2fbbc 100644
--- a/ui/src/components/DiffRelease/diff.worker.js
+++ b/ui/src/components/DiffRelease/diff.worker.js
@@ -2,14 +2,12 @@ import { createTwoFilesPatch } from 'diff';
 
 onmessage = e => {
   const [
-    firstFilename,
-    secondFilename,
     firstReleaseString,
     secondReleaseString,
   ] = e.data;
   const releaseDiff = createTwoFilesPatch(
-    firstFilename,
-    secondFilename,
+    'Revision (left)',
+    'Revision (right)',
     firstReleaseString,
     secondReleaseString
   );
diff --git a/ui/src/components/DiffRelease/index.jsx b/ui/src/components/DiffRelease/index.jsx
index 7eb4ec47..dc3de26b 100644
--- a/ui/src/components/DiffRelease/index.jsx
+++ b/ui/src/components/DiffRelease/index.jsx
@@ -49,8 +49,6 @@ function DiffRelease(props) {
   const {
     firstRelease,
     secondRelease,
-    firstFilename,
-    secondFilename,
     className,
   } = props;
   const classes = useStyles();
@@ -79,12 +77,10 @@ function DiffRelease(props) {
 
   useEffect(() => {
     diffWorker.postMessage([
-      firstFilename,
-      secondFilename,
       JSON.stringify(deepSortObject(firstRelease), null, 2),
       JSON.stringify(deepSortObject(secondRelease), null, 2),
     ]);
-  }, [firstFilename, secondFilename, firstRelease, secondRelease]);
+  }, [firstRelease, secondRelease]);
 
   const handleRowRender = ({ index, key, style }) => {
     const line = releaseLinesDiff[index];
@@ -139,8 +135,6 @@ function DiffRelease(props) {
 DiffRelease.propTypes = {
   firstRelease: object.isRequired,
   secondRelease: object.isRequired,
-  firstFilename: string.isRequired,
-  secondFilename: string.isRequired,
   className: string,
 };
```

Reply:
Hmmmm, I think that will help but it may not entirely fix the issue? We'll still get two renders, because `firstRelease` and `secondRelease` may both change, so if the first worker finishes after the second one, that will be wrong. I could be misunderstanding though.

Reply:
Ah, I think I misunderstood the issue. Correct me if I'm wrong, but you're looking for a way to not use the result of the worker if that diff belongs to a different pair of releases. If that's what you want to do, then maybe (let me know if it helps) we could either:
* add a cleanup function to useEffect that would terminate any ongoing worker before that useEffect runs again. (preferred)

```jsx
useEffect(() => {
 // ...
 return function cleanup() {
    diffWorker.terminate();
  };
});
```

* append another property (e.g., revision date) to the worker result and then check that it belongs to what the user selected and if it matches.

Reply:
I didn't know about cleanup functions - that's really cool! I suspect there may be a tiny possibility of a race condition still (if it completes before the termination does), but it'll probably be perfectly fine for our use case, especially since I've been unable to repro anywhere but my laptop.

Thanks!

Reply:
After some additional reading, I think I'm wrong about the race condition.